File: cpp/1082.cpp
#include <iostream>
#include <unordered_map>
#include <vector>
#include <stack>
#include <algorithm>

using namespace std;

class Graph
{
    public:
        int vertices;
        unordered_map<int, vector<int>> graph;

    void add_edge(int a, int b)
    {
        graph[a].push_back(b);
        graph[b].push_back(a);
    }

    void get_cycles(int vertex, bool visited[], vector<int>& connected)
    {
        visited[vertex] = true;
        connected.push_back(vertex);
        for (auto i: graph[vertex]) {
            if (!visited[i]) {
                get_cycles(i, visited, connected);
            }
        }
    }

    Graph get_transpose()
    {
        Graph new_graph;
        new_graph.vertices = vertices;
        for (auto i: graph) {
            for (auto j: i.second) {
                new_graph.add_edge(j, i.first);
            }
        }
        return new_graph;
    }

    void fill_stack(int vertex, bool visited[], stack<int>& stack)
    {
        visited[vertex] = true;
        for (auto i: graph[vertex]) {
            if (!visited[i]) {
                fill_stack(i, visited, stack);
            }
        }
        stack.push(vertex);
    }

    void get_sccs(vector<vector<int>>& result)
    {
        stack<int> stack;
        bool visited[vertices] = {false};
        for (int i = 0; i < vertices; i++) {
            if (!visited[i]) {
                fill_stack(i, visited, stack);
            }
        }
        Graph gr_transpose;
        gr_transpose = get_transpose();
        bool t_visited[vertices] = {false};
        int i;
        while (!stack.empty()) {
            i = stack.top();
            stack.pop();
            if (!t_visited[i]) {
                vector<int> connected;
                gr_transpose.get_cycles(i, t_visited, connected);
                sort(connected.begin(), connected.end());
                result.push_back(connected);
            }
        }
    }
};

int main()
{
    int n, count, vertices, connections;
    char a, b;
    cin >> n;
    for (count = 0; count < n; count++) {
        cin >> vertices >> connections;
        Graph gr;
        gr.vertices = vertices;
        for (connections; connections > 0; connections--) {
            cin >> a >> b;
            gr.add_edge((int)(a) - 97, int(b) - 97);
        }
        vector<vector<int>> result;
        gr.get_sccs(result);
        sort(result.begin(), result.end());
        cout << "Case #" << count + 1 << ":\n";
        for (auto i: result) {
            for (auto j: i) {
                cout << (char)(j + 97) << ',';
            }
            cout << '\n';
        }
        cout << result.size() << " connected components\n\n";
    }
    return 0;
}

File: py/1082.py
from sys import stdin


class Graph:
    def __init__(self, vertices) -> None:
        self.vertices = vertices
        self.graph = {i: set() for i in range(0, vertices + 1)}

    def add_edge(self, a, b) -> None:
        self.graph[a].add(b)
        self.graph[b].add(a)

    def get_cycles(self, vertice, visited, connected) -> None:
        visited[vertice] = True
        connected.append(vertice)
        for i in self.graph[vertice]:
            if not visited[i]:
                self.get_cycles(i, visited, connected)

    def get_transpose(self) -> object:
        graph = Graph(self.vertices)
        for i in self.graph:
            for j in self.graph[i]:
                graph.add_edge(j, i)
        return graph

    def fill_stack(self, vertice, visited, stack) -> None:
        visited[vertice] = True
        for i in self.graph[vertice]:
            if not visited[i]:
                self.fill_stack(i, visited, stack)
        stack.append(vertice)

    def get_sccs(self) -> list:
        stack = []
        visited = [False] * self.vertices
        for i in range(self.vertices):
            if not visited[i]:
                self.fill_stack(i, visited, stack)
        gr_transpose = self.get_transpose()
        visited = [False] * self.vertices
        result = []
        while stack:
            i = stack.pop()
            if not visited[i]:
                connected = []
                gr_transpose.get_cycles(i, visited, connected)
                result.append(connected)
        return result


def main() -> None:
    input = stdin.readline
    n = int(input())
    for case in range(n):
        vertices, connections = [int(i) for i in input().split()]
        gr = Graph(vertices)
        for j in range(connections):
            a, b = [int(ord(i) - 97) for i in input().split()]
            gr.add_edge(a, b)
        result = gr.get_sccs()
        result = sorted([sorted(i) for i in result])
        result = [[chr(i + 97) for i in j] for j in result]
        print(f'Case #{case + 1}:')
        for i in result:
            print(*i, '', sep=',')
        print(f'{len(result)} connected components\n')


if __name__ == '__main__':
    main()
